### DISCUSSION TITLE
Open a window on a specific workspace at start once

### DISCUSSION THREAD

**Discussion by boxdot** [hyprwm/Hyprland#982] [Q&A]
Is there a way to open a specific window on a predefined workspace when starting Hyprland. But after that to allow this window on any workspace? 

I would like to start firefox on a predefined workspace on init. But after the start, it should be possible for firefox to be open on any other workspace too.

**Comment by vaxerski** [comment]
script it

**Comment by Lorde-Soup** [comment]
https://wiki.hyprland.org/FAQ/#how-do-i-autostart-my-favorite-apps

**Reply by isshin1** [reply]
> https://wiki.hyprland.org/FAQ/#how-do-i-autostart-my-favorite-apps

is there any way i can combine it with hot reload, this doesnt seem to work

```
exec = pgrep nautilus || [workspace special:nautilus silent] nautilus 

```

**Comment by mdvmeijer** [comment]
Instead of a cleanup-after-start script as stated in the FAQ, you can also simply add a workspace rule to exec-once like this:
`exec-once = [workspace 9 silent] alacritty`

Additional instances of alacritty that you open yourself, will simply open on the focused workspace.

**Comment by samEngineering** [comment]
Anyone having trouble doing this on the special workspace?
`exec-once = [workspace special silent] kitty`
Nothing in @Lord-Soup's link about that.

**Reply by MightyPlaza** [reply]
should just work
make sure to get after v0.30.0, I'm pretty sure this section was rewritten since last release

**Reply by samEngineering** [reply]
Turns out it is not a `hyprland` issue. launching other programs in this fashion works as expected. It is specific to `kitty`. Namely, `[glfw error 65544]`

**Reply by hyperdotbat** [reply]
well thats weird, because it also doesnt work [silently] for me; with Discord and Spotify
EDIT: actually all workspaces even regular ones focus the apps, this is probably caused by setting ` focus_on_activate ` but it shouldnt

**Comment by LouDou** [comment]
I'm trying to attempt this too, but with the following lines:

```
exec-once = [workspace 1 silent] slack
exec-once = [workspace 1 silent] /opt/google/chrome/google-chrome --profile-directory=Default --app-id=cifhbcnohmdccbgoicgdjpfamggdegmo
exec-once = [workspace 2 silent] /opt/google/chrome/google-chrome
exec-once = [workspace 3 silent] code
exec-once = [workspace 3 silent] terminology
exec-once = [workspace 8 silent] qjackctl
exec-once = [workspace 9 silent] /home/[REDACTED]/bin/smartgit/smartgit-20_2_3/bin/smartgit.sh
exec-once = [workspace 10 silent] firefox
```

All the apps seem to start on workspace 1 when I log in. Is there some other config which needs to be set or cleared to make this work?

```
❯ Hyprland --version
Hyprland, built from branch  at commit 4520b30d498daca8079365bdb909a8dea38e8d55  (version: bump to 0.44.1).
Date: Wed Oct 9 12:54:39 2024
Tag: v0.44.1, commits: 5309
built against aquamarine 0.4.2


flags: (if any)
```

**Reply by isshin1** [reply]
> I'm trying to attempt this too, but with the following lines:
> 
> ```
> exec-once = [workspace 1 silent] slack
> exec-once = [workspace 1 silent] /opt/google/chrome/google-chrome --profile-directory=Default --app-id=cifhbcnohmdccbgoicgdjpfamggdegmo
> exec-once = [workspace 2 silent] /opt/google/chrome/google-chrome
> exec-once = [workspace 3 silent] code
> exec-once = [workspace 3 silent] terminology
> exec-once = [workspace 8 silent] qjackctl
> exec-once = [workspace 9 silent] /home/[REDACTED]/bin/smartgit/smartgit-20_2_3/bin/smartgit.sh
> exec-once = [workspace 10 silent] firefox
> ```
> 
> All the apps seem to start on workspace 1 when I log in. Is there some other config which needs to be set or cleared to make this work?
> 
> ```
> ❯ Hyprland --version
> Hyprland, built from branch  at commit 4520b30d498daca8079365bdb909a8dea38e8d55  (version: bump to 0.44.1).
> Date: Wed Oct 9 12:54:39 2024
> Tag: v0.44.1, commits: 5309
> built against aquamarine 0.4.2
> 
> 
> flags: (if any)
> ```

most are electron, hyprland only keep track of primary pid, but electron apps generally create a child process and spawn new window using that . so wont work

**Reply by LouDou** [reply]
That's a real shame. Is there any hope of this ever working?

**Reply by mlpeter33** [reply]
have u tried using a windowrule? like forcing an app to only spawn on a specific workspace:
 `windowrulev2 = workspace 2, class:^(Firefox)$` I haven't test it but maybe it can work

**Reply by isshin1** [reply]
> have u tried using a windowrule? like forcing an app to only spawn on a specific workspace: `windowrulev2 = workspace 2, class:^(Firefox)$` I haven't test it but maybe it can work

windowrules are permanent , we need one time, preferably on launch

**Comment by bassicallychris** [comment]
I'm having the same struggle when working with Firefox, my use case is similar to @LouDou where I'd like to open profiles mapped to specific workspaces. I'm using `gtk-launch` to use desktop files that target these specific profiles. 
```shell
exec-once = gtk-launch firefox_profile_1
exec-once = gtk-launch firefox_profile_2
exec-once = gtk-launch firefox_profile_3
exec-once = gtk-launch firefox_profile_4
```
I've attempted using 
```shell
exec-once = [workspace 2 silent] gtk-launch firefox_profile_1
```
but it only launches on the currently active workspace. (testing using `exec` instead of `exec-once`)
Help would be greatly appreciated here.

**Reply by yeldiRium** [reply]
This is due to this line in the docs: `It records the PID of the spawned process and uses that. For example, if your process forks and then the fork opens a window, this will not work.`

I assume `gtk-launch` does just that. I also ran into this problem today and would really love a solution for this. Someone suggested having [exec-desktop and exec-once-desktop](https://github.com/hyprwm/Hyprland/issues/2912) to launch .desktop files, which would be great for my use case, but was closed as not planned.

**Reply by Roxxas96** [reply]
Hey there ! I am running on the same issue, have some one found a workaround ? I tried running `exec-once` with `hyprctl dispatch workspace` but it is running into the same issue, ex:
```
exec-once = hyprctl dispatch workspace 2; \
            gtk-launch code
```

**Reply by AlexAntonik** [reply]
If you don’t need it to start right at system launch and are okay with the ```code``` starting on workspace creation, then ```workspace = 2, on-created-empty:code``` should work.

**Reply by Roxxas96** [reply]
Thanks ! I just tried the workaround and although it is not exactly reproducing exec-once as it will start your app each time you enter your workspace when it's empty, it is an acceptable behavior for my use case.

**Comment by AlexAntonik** [comment]
Searched for solution to this - nothing found.
It would be nice to have some flag for this to work by matching window name or class.

Like this:
> ```
>exec-once = [workspace 1 silent match-name] slack
>exec-once = [workspace 1 silent match-name] /opt/google/chrome/google-chrome --profile-directory=Default --app-
>exec-once = [workspace 2 silent match-class] /opt/google/chrome/google-chrome
>exec-once = [workspace 3 silent match-class] code
>````

**Comment by JonKoala** [comment]
The only way I found to make this work with apps that spawn child processes to create their windows was to temporarily set a `windowrule`, then unset it after a short delay.

```
windowrule = match:class code, workspace 2 silent
exec-once = code
exec-once = sleep 5 && hyprctl keyword windowrule "match:class code, workspace unset"
```

Not the cleanest solution, but it works for my use cases.

**Reply by hyperdotbat** [reply]
fyi now windowrules have been given a new v3 syntax, but this ultimately should be an even better thing because each rule has a name
I think your solution is good, Ive just been using static windowrules themselves and kind of dedicating workspaces for certain things

**Comment by trueBluewolf** [comment]
Hey had the same issue but I got it to work. I was looking to maybe see if workspace 2 exist at the start and it turns out it doesn't yet until you manually switch to it or start an app there. So just make sure you create workspace 2 on start then launch app with the [workspace #].
```
workspace = 2, persistent, monitor: YOUR MONITOR NAME HERE
exec-once = [workspace 2] APP HERE
```

**Comment by dwolrdcojp** [comment]
This syntax used to work in the config just fine for starting apps on launch 

`exec-once = [workspace 2 silent] chromium`

it is now broken in the latest update... what gives?

**Reply by MartinLoeper** [reply]
happy to see i am not the only one experiencing this... @vaxerski do you have any idea why that might be? ;)
I'm using e.g.: `exec-once = [workspace 5 silent] uwsm app -- vivaldi`

**Reply by Dregu** [reply]
The new environment token based exec rule system is pretty neat and works with many more forking child process things than the old pid system, but chrome flavored browsers specifically wipe the whole process environment for security reasons, so hl can't find the exec rule token. We'd need to combine the two to make the stupid browsers work...
